src/compiler/core.ts

```diff
@@ -2133,4 +2146,74 @@ namespace ts {
     export function fill<T>(length: number, cb: (index: number) => T): T[] {
         return new Array(length).fill(0).map((_, i) => cb(i));
     }
+
+    /**
+     * A list of sorted and unique values. Optimized for best performance when items are added
+     * in the sort order.
+     */
+    export class SortedUniqueList<T> implements Push<T> {
+        private relationalComparer: (a: T, b: T) => Comparison;
+        private equalityComparer: (a: T, b: T) => boolean;
+        private sortedAndUnique = true;
+        private copyOnWrite = false;
+        private unsafeArray: T[] = [];
+        private unsafeLast: T | undefined = undefined;
+
+        constructor(relationalComparer: Comparer<T>, equalityComparer: EqualityComparer<T>) {

```

Comment:
> [...] should I assume that the values have only a partial ordering? Otherwise having both `relationalComparer` and `equalityComparer` is redundant?

In the case of its current usage, yes. The relational comparison is used to compare source positions only, but the equality comparer used is more detailed. I can change the comparison to cover the equalityComparer